src/input.rs

```diff
@@ -1081,6 +1311,83 @@ where
     go_extra!();
 }
 
+pub struct Foldr<P, F, A, B, E = (), S = ()> {
+    parser: P,
+    folder: F,
+    phantom: PhantomData<(A, B, E, S)>,
+}
+
+impl<P: Copy, F: Copy, A, B, E, S> Copy for Foldr<P, F, A, B, E, S> {}
+impl<P: Clone, F: Clone, A, B, E, S> Clone for Foldr<P, F, A, B, E, S> {
+    fn clone(&self) -> Self {
+        Foldr {
+            parser: self.parser.clone(),
+            folder: self.folder.clone(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, I, P, F, A, B, E, S> Parser<'a, I, E, S> for Foldr<P, F, A, B, E, S>
+where
+    I: Input + ?Sized,
+    P: Parser<'a, I, E, S, Output = (A, B)>,
+    E: Error<I::Token>,
+    S: 'a,
+    A: IntoIterator,
+    A::IntoIter: DoubleEndedIterator,
+    F: Fn(A::Item, B) -> B,
+{
+    type Output = B;
+
+    fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, S>) -> PResult<M, Self::Output, E> where Self: Sized {
+        self.parser.go::<M>(inp)
+            .map(|out|
+                M::map(out, |(init, end)| init.into_iter().rev().fold(end, |b, a| (self.folder)(a, b)))
```

Comment:
I wonder whether a dedicated combinator for this would be better eventually, avoiding the extra `Vec`. Something like `lhs.fold(rhs, |lhs. rhs| ...)` and `lhs.fold_into(rhs, |lhs, rhs| ...)`...